src/python/turicreate/_deps/__init__.py

```diff
@@ -9,59 +9,346 @@
+    if six.PY2:
+        # imp importing submodule is too tricky
+        # moulde and pkg naming should follow pep8
+        if not all([x.isalnum() or x in '._' for x in name]):
+            raise ValueError('invalid module name')
+        exec("import %s as _py2_ret" % name)
+        return _py2_ret
+    else:
+        return importlib.import_module(name)
+
+class LazyModuleLoader(ModuleType, object):
+    """ defer to load a module. If it's loaded, no exception but warning will be given. """
+    # read-only member list -> used by __setattr__; see below
+    _my_attrs_ = ['is_loaded', 'reload', 'get_module',
+                  '_name', '_init_func', '_module', '_loaded']
+    # ensure singleton; must be guarded by _ImportLockContext lock
+    registry_ = set()
+
+    def __init__(self, name, init_func=default_init_func, is_lambda=False):
+        """
+        Once write then read only. `reload` can reset the state.
+        Only singleton instance is allowed.
+        Context _ImportLockContext is reentrant.
+
+        This Class only supports syntax `import a.b [as c]`;
+        you need to translate `from a import b as c` to the form described above.
+
+        LazyModuleLoader hijacks sys.modules[name] to defer the
+        use of the module if the module is not already loaded yet.
+
+        The load of module is deferred until LazyModuleLoader is called
+        by __getattr__, __setattr__, or __delattr__ on `attr` other than
+        `dir(LazyModuleLoader())` The side effect is that sys.modules[name]
+        will be set to real module object.
+
+        >>> np = LazyModuleLoader('numpy')
+        >>> import numpy # won't actually load
+        >>> np.ndarray # load the numpy module
+
+        If the module is loaded before LazyModuleLoader loads the underlying module,
+        then LzyModuleLoader won't do anything but just a thin wrapper to forward
+        any requests to the real module object.
+
+        >>> import numpy # load the numpy
+        >>> np = LazyModuleLoader('numpy') # no effect
+        >>> np.ndarray # works as the same as numpy.ndarray
+
+        This example is very nasty since it will actualy delete
+        the LazyModuleLoader object due to garbage collection:
+
+        >>> def baz():
+        ...     LazyModuleLoader('foo').bar
+        >>> baz() # the actual module `foo` is loaded
+
+        """
+        if init_func is None:
+            raise ValueError("init function should not be None")
+        if not isinstance(name, six.string_types):
+            raise ValueError("name must be str type")
+        if len(name) == 0:
+            raise ValueError("module name should not be empty")
+        if name.startswith('.'):
+            raise ValueError("only support absolute path import style")
+        # must set first to avoid recursion
+        self._name = name
+        self._init_func = init_func
+
+        with _ImportLockContext():
```

Comment:
I still don't understand why an import lock is needed.

Reply:
the importlib has it. I guess probably the import is done in C thread.